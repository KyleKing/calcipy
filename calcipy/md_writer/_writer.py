"""Markdown Machine."""

import json
import re
from pathlib import Path

import pandas as pd
from beartype import beartype
from beartype.typing import Any, Callable, Dict, List, Optional
from corallium.file_helpers import read_lines
from corallium.log import logger
from transitions import Machine

from ..file_search import find_project_files_by_suffix
from ..invoke_helpers import get_project_path

HandlerLookupT = Dict[str, Callable[[str, Path], List[str]]]
"""Handler Lookup."""


class _ParseSkipError(RuntimeError):
    """Exception caught if the handler does not want to replace the text."""


class _ReplacementMachine(Machine):
    """State machine to replace content with user-specified handlers.

    Uses `{cts}` and `{cte}` to demarcate sections (short for calcipy_template start|end)

    """

    def __init__(self) -> None:
        """Initialize the state machine."""
        self.state_user = 'user'
        self.state_auto = 'autoformatted'
        transitions = [
            {
                'trigger': 'start_auto',
                'source': self.state_user,
                'dest': self.state_auto,
            }, {
                'trigger': 'end',
                'source': self.state_auto,
                'dest': self.state_user,
            },
        ]
        super().__init__(
            states=[self.state_user, self.state_auto],
            initial=self.state_user,
            transitions=transitions,
        )

    @beartype
    def parse(
        self, lines: List[str], handler_lookup: HandlerLookupT,
        path_file: Path,
    ) -> List[str]:
        """Parse lines and insert new_text based on provided handler_lookup.

        Args:
            lines: list of string from source file
            handler_lookup: Lookup dictionary for autoformatted sections
            path_file: optional path to the file. Only useful for debugging

        Returns:
            List[str]: modified list of strings

        """
        updated_lines = []
        for line in lines:
            updated_lines.extend(self._parse_line(line, handler_lookup, path_file))
        return updated_lines

    @beartype
    def _parse_line(
        self, line: str, handler_lookup: HandlerLookupT, path_file: Path,
    ) -> List[str]:
        """Parse lines and insert new_text based on provided handler_lookup.

        Args:
            line: single line
            handler_lookup: Lookup dictionary for autoformatted sections
            path_file: optional path to the file. Only useful for debugging

        Returns:
            List[str]: modified list of strings

        """
        lines: List[str] = []
        if '{cte}' in line and self.state == self.state_auto:  # end
            self.end()
        elif '{cts}' in line:  # start
            self.start_auto()
            matches = [text_match for text_match in handler_lookup if text_match in line]
            if len(matches) == 1:
                try:
                    lines.extend(handler_lookup[matches[0]](line, path_file))
                except _ParseSkipError:
                    lines.append(line)
                    self.end()
            else:
                logger.warning('Could not parse. Skipping:', line=line)
                lines.append(line)
                self.end()
        elif self.state == self.state_user:
            lines.append(line)
        # else: discard the lines in the auto-section
        return lines


_VAR_COMMENT_HTML = r'<!-- {cts} (?P<key>[^=]+)=(?P<value>[^;]+);'
"""Regex for extracting the variable from an HTML code comment."""


@beartype
def _parse_var_comment(line: str, matcher: str = _VAR_COMMENT_HTML) -> Dict[str, str]:
    """Parse the variable from a matching comment.

    Args:
        line: string from source file
        matcher: string regex pattern to match. Default is `_RE_VAR_COMMENT_HTML`

    Returns:
        Dict[str, str]: single key and value pair based on the parsed comment

    """
    if match := re.compile(matcher).match(line.strip()):
        matches = match.groupdict()
        return {matches['key']: matches['value']}
    return {}


@beartype
def _handle_source_file(line: str, path_file: Path) -> List[str]:
    """Replace commented sections in README with linked file contents.

    Args:
        line: first line of the section
        path_file: path to the file that contained the string

    Returns:
        List[str]: list of auto-formatted text

    """
    key, path_rel = [*_parse_var_comment(line).items()][0]
    path_base = get_project_path() if path_rel.startswith('/') else path_file.resolve().parent
    path_source = path_base / path_rel.lstrip('/')
    language = path_source.suffix.lstrip('.')
    lines_source = [f'```{language}', *read_lines(path_source), '```']
    if not path_source.is_file():
        logger.warning('Could not locate source file', path_source=path_source)

    line_start = f'<!-- {{cts}} {key}={path_rel}; -->'
    line_end = '<!-- {cte} -->'
    return [line_start, *lines_source, line_end]


@beartype
def _format_cov_table(coverage_data: Dict[str, Any]) -> List[str]:
    """Format code coverage data table as markdown.

    Args:
        coverage_data: dictionary created by `python -m coverage json`

    Returns:
        List[str]: list of string lines to insert

    """
    col_key_map = {
        'Statements': 'num_statements',
        'Missing': 'missing_lines',
        'Excluded': 'excluded_lines',
        'Coverage': 'percent_covered',
    }
    records = [
        {
            'File': f'`{Path(path_file).as_posix()}`',
            **{col: file_obj['summary'][key] for col, key in col_key_map.items()},
        } for path_file, file_obj in coverage_data['files'].items()
    ]
    records.append({
        'File': '**Totals**',
        **{col: coverage_data['totals'][key] for col, key in col_key_map.items()},
    })
    # Format table for Github Markdown
    df_cov = pd.DataFrame(records)
    df_cov['Coverage'] = df_cov['Coverage'].round(1).astype(str) + '%'
    lines_table = str(df_cov.to_markdown(index=False, tablefmt='github')).split('\n')
    short_date = coverage_data['meta']['timestamp'].split('T')[0]
    lines_table.extend(['', f'Generated on: {short_date}'])
    return lines_table


@beartype
def _handle_coverage(line: str, _path_file: Path, path_coverage: Optional[Path] = None) -> List[str]:
    """Read the coverage.json file and write a Markdown table to the README file.

    Args:
        line: first line of the section
        path_file: path to the file that contained the string

    Returns:
        List[str]: list of auto-formatted text

    Raises:
        _ParseSkipError: if the "coverage.json" file is not available

    """
    path_coverage = path_coverage or get_project_path() / 'coverage.json'
    if not path_coverage.is_file():
        msg = f'Could not locate: {path_coverage}'
        raise _ParseSkipError(msg)
    coverage_data = json.loads(path_coverage.read_text())
    lines_cov = _format_cov_table(coverage_data)
    line_end = '<!-- {cte} -->'
    return [line, *lines_cov, line_end]


@beartype
def write_autoformatted_md_sections(
    handler_lookup: Optional[HandlerLookupT] = None,
    paths_md: Optional[List[Path]] = None,
) -> None:
    """Populate the auto-formatted sections of markdown files with user-configured logic."""
    _lookup: HandlerLookupT = handler_lookup or {
        'COVERAGE ': _handle_coverage,
        'SOURCE_FILE=': _handle_source_file,
    }

    paths = paths_md or find_project_files_by_suffix(get_project_path()).get('md') or []
    for path_md in paths:
        logger.text_debug('Processing', path_md=path_md)
        if md_lines := _ReplacementMachine().parse(read_lines(path_md), _lookup, path_md):
            path_md.write_text('\n'.join(md_lines) + '\n', encoding='utf-8')
