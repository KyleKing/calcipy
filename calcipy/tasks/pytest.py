"""TBD CLI."""

from pathlib import Path
from beartype.typing import Dict, List, Tuple, Optional, Callable
from functools import partial
from invoke import task, Context
from beartype import beartype
import logging
from shoal import get_logger
from shoal._log import configure_logger

logger = get_logger()

@beartype
def _inner_task(ctx: Context, *, cli_args: List[str]) -> None:
    """Shared task logic."""
    gto = ctx.config.gto
    configure_logger(log_level={3: logging.NOTSET, 2: logging.DEBUG, 1: logging.INFO, 0: logging.WARNING}.get(gto.verbose) or logging.ERROR)

    ctx.run(
        f'poetry run python -m pytest ./tests {" ".join(cli_args)}',
        # FYI: see ../tasks/nox.py for open questions
        echo=True, pty=True,
    )


@task(
    default=True,
    help={},
)
@beartype
def default(ctx: Context) -> None:
    """TBD."""
    _inner_task(ctx, cli_args=[])


# all: " --ff -vv"
# need flags for marker and keyword
# Coverage (sequence of commands)
