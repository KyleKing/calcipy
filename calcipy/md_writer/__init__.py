try:
    from ._writer import write_autoformatted_md_sections
except ImportError as exc:
    raise RuntimeError("The 'calcipy[doc]' extras are missing") from exc
