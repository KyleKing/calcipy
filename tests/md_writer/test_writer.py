"""Test doit_tasks/doc.py."""

import json
import shutil
from functools import partial
from pathlib import Path

import pytest
from beartype import beartype
from beartype.typing import List

from calcipy.md_writer._writer import (
    _format_cov_table,
    _handle_coverage,
    _handle_source_file,
    _parse_var_comment,
    write_autoformatted_md_sections,
)

from ..configuration import TEST_DATA_DIR

SAMPLE_README_PATH = TEST_DATA_DIR / 'sample_doc_files' / 'README.md'

_COVERAGE_SAMPLE_DATA = {
    'meta': {'timestamp': '2021-06-03T19:37:11.980123'},
    'files': {
        'calcipy/doit_tasks/base.py': {
            'summary': {
                'covered_lines': 20,
                'num_statements': 22,
                'percent_covered': 90.9090909090909,
                'missing_lines': 2,
                'excluded_lines': 3,
            },
        },
        'calcipy/doit_tasks/code_tags.py': {
            'summary': {
                'covered_lines': 31,
                'num_statements': 75,
                'percent_covered': 41.333333333333336,
                'missing_lines': 44,
                'excluded_lines': 0,
            },
        },
    },
    'totals': {
        'covered_lines': 51,
        'num_statements': 97,
        'percent_covered': 52.57732,
        'missing_lines': 46,
        'excluded_lines': 3,
    },
}
"""Sample coverage data generated with `python -m coverage json`."""


def test_format_cov_table():
    result = _format_cov_table(_COVERAGE_SAMPLE_DATA)

    assert result == [
        '| File                              |   Statements |   Missing |   Excluded | Coverage   |',
        '|-----------------------------------|--------------|-----------|------------|------------|',
        '| `calcipy/doit_tasks/base.py`      |           22 |         2 |          3 | 90.9%      |',
        '| `calcipy/doit_tasks/code_tags.py` |           75 |        44 |          0 | 41.3%      |',
        '| **Totals**                        |           97 |        46 |          3 | 52.6%      |',
        '',
        'Generated on: 2021-06-03',
    ]


def test_write_autoformatted_md_sections(fix_test_cache):
    path_new_readme = fix_test_cache / SAMPLE_README_PATH.name
    shutil.copyfile(SAMPLE_README_PATH, path_new_readme)
    path_cover = fix_test_cache / 'coverage.json'
    path_cover.write_text(json.dumps(_COVERAGE_SAMPLE_DATA))
    placeholder = '<!-- {cts} SOURCE_FILE_TEST=/tests/conftest.py; -->\n<!-- {cte} -->'
    was_placeholder = placeholder in path_new_readme.read_text()

    write_autoformatted_md_sections(handler_lookup={
        'SOURCE_FILE_TEST': _handle_source_file,
        'COVERAGE_TEST': partial(_handle_coverage, path_coverage=path_cover),
    }, paths_md=[path_new_readme])

    text = path_new_readme.read_text()
    assert was_placeholder
    assert placeholder not in text
    assert '<!-- {cts} COVERAGE_TEST -->\n| File    ' in text
    assert '| File                              |   Statements |   Missing |   Excluded | Coverage   |' in text
    assert '| `calcipy/doit_tasks/base.py`      |           22 |         2 |          3 | 90.9%      |' in text
    assert '| **Totals**                        |           97 |        46 |          3 | 52.6%      |' in text
    assert 'Generated on: 2021-06-03' in text
    assert '<!-- {cts} SOURCE_FILE_TEST=/tests/conftest.py; -->\n```py\n"""PyTest configuration.' in text


@pytest.mark.parametrize(
    ('line', 'match'), [
        ('<!-- {cts} rating=1; (User can specify rating on scale of 1-5) -->', {'rating': '1'}),
        ('<!-- {cts} path_image=imgs/image_filename.png; -->', {'path_image': 'imgs/image_filename.png'}),
        ('<!-- {cts} tricky_var_3=-11e-21; -->', {'tricky_var_3': '-11e-21'}),
    ],
)
def test_parse_var_comment(line, match):
    result = _parse_var_comment(line)

    assert result == match


@beartype
def _star_parser(line: str, path_md: Path) -> List[str]:  # noqa: ARG001
    rating = int(_parse_var_comment(line)['rating'])
    return [f'RATING={rating}']


def test_write_autoformatted_md_sections_custom(fix_test_cache):
    path_new_readme = fix_test_cache / SAMPLE_README_PATH.name
    shutil.copyfile(SAMPLE_README_PATH, path_new_readme)

    write_autoformatted_md_sections(handler_lookup={'rating': _star_parser}, paths_md=[path_new_readme])

    text = path_new_readme.read_text()
    assert '\n<!-- {cts} rating=' in SAMPLE_README_PATH.read_text()
    assert '\n<!-- {cts} rating=' not in text
    assert '\n\nRATING=4\n\n' in text
    assert """<!-- {cts} name_image=NA.png; (User can specify image name) -->
<!-- Capture image -->
<!-- {cte} -->""" in text
