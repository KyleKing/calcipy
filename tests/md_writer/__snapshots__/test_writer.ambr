# serializer version: 1
# name: test_format_cov_table
  '''
  | File                              | Statements | Missing | Excluded | Coverage |
  |-----------------------------------|-----------:|--------:|---------:|---------:|
  | `calcipy/doit_tasks/base.py`      | 22         | 2       | 3        | 90.9%    |
  | `calcipy/doit_tasks/code_tags.py` | 75         | 44      | 0        | 41.3%    |
  | **Totals**                        | 97         | 46      | 3        | 52.6%    |

  Generated on: 2021-06-03
  '''
# ---
# name: test_write_template_formatted_md_sections
  '''
  <!-- Do not modify sections within "{ct*}". They are updated when building the docs -->

  # Title

  > Test markdown formatting from [recipes](https://github.com/KyleKing/recipes)

  <!-- {cts} rating=4; (User can specify rating on scale of 1-5) -->
  Personal rating: *Not yet rated*
  <!-- {cte} -->

  <!-- {cts} name_image=NA.png; (User can specify image name) -->
  <!-- Capture image -->
  <!-- {cte} -->

  <!-- {cts} COVERAGE_TEST -->
  | File                              | Statements | Missing | Excluded | Coverage |
  |-----------------------------------|-----------:|--------:|---------:|---------:|
  | `calcipy/doit_tasks/base.py`      | 22         | 2       | 3        | 90.9%    |
  | `calcipy/doit_tasks/code_tags.py` | 75         | 44      | 0        | 41.3%    |
  | **Totals**                        | 97         | 46      | 3        | 52.6%    |

  Generated on: 2021-06-03
  <!-- {cte} -->

  <!-- {cts} SOURCE_FILE_TEST=/tests/test_zz_calcipy.py; -->
  ```py
  """Final test alphabetically (zz) to catch general integration cases."""

  from pathlib import Path

  from corallium.tomllib import tomllib

  from calcipy import __version__


  def test_version():
      """Check that PyProject and package __version__ are equivalent."""
      data = Path('pyproject.toml').read_text(encoding='utf-8')

      result = tomllib.loads(data)['project']['version']

      assert result == __version__
  ```
  <!-- {cte} -->

  '''
# ---
